espymetrics/dal/sqlite_dal.py

```diff
@@ -13,25 +13,116 @@
-        self.file_path = file_path
+        self.__file_path = file_path
+
+        conn = sqlite3.connect(file_path)
+        conn.execute(CREATE_TABLE)
 
     def add_record(self, in_dict):
-        pass
+        query = ('INSERT INTO ' + TABLE_NAME +
+                 ' (hostfqdn, hostname, python$argv, python$bin, ' +
+                 'python$build_date, python$compiler, python$have_readline, ' +
+                 'python$have_ucs4, python$is_64bit, python$version, python$version_full, ' +
+                 'time$utc_epoch, time$std_utc_offset, ' +
+                 'uname$0, uname$1, uname$2, uname$3, uname$4, uname$5, ' +
+                 'username, uuid) VALUES (')
+        query += '"' + in_dict['hostfqdn'] + '", '
+        query += '"' + in_dict['hostname'] + '", '
+        query += '"' + in_dict['python']['argv'] + '", '
+        query += '"' + in_dict['python']['bin'] + '", '
+        query += '"' + in_dict['python']['build_date'] + '", '
+        query += '"' + in_dict['python']['compiler'] + '", '
+        query += '"' + str(in_dict['python']['have_readline'])[0] + '", '
+        query += '"' + str(in_dict['python']['have_ucs4'])[0] + '", '
+        query += '"' + str(in_dict['python']['is_64bit'])[0] + '", '
+        query += '"' + in_dict['python']['version'] + '", '
+        query += '"' + in_dict['python']['version_full'] + '", '
+        query += '"' + str(in_dict['time']['utc_epoch']) + '", '
+        query += '"' + str(in_dict['time']['std_utc_offset']) + '", '
+        query += '"' + in_dict['uname'][0] + '", '
+        query += '"' + in_dict['uname'][1] + '", '
+        query += '"' + in_dict['uname'][2] + '", '
+        query += '"' + in_dict['uname'][3] + '", '
+        query += '"' + in_dict['uname'][4] + '", '
+        query += '"' + in_dict['uname'][5] + '", '
+        query += '"' + in_dict['username'] + '", '
+        query += '"' + in_dict['uuid']
+        query += '")'
+
+        try:
+            conn = sqlite3.connect(self.__file_path)
+            conn.isolation_level = None  # autocommit
+            conn.execute(query)
+            conn.close()
+        except:
+            pass
+
+        global total_count
+        total_count += 1
+        return
 
     def raw_query(self, query):
-        pass
+        try:
+            conn = sqlite3.connect(self.__file_path)
+            rows = conn.execute(query).fetchall()
+            return rows
+        except:
+            return -1
 
     def select_records(self, limit=None, group_by=None):
-        pass
+        ret = {}
+        if not group_by:
+            return ret
+        group_by = group_by.replace('.', '$')
+        query = 'SELECT ROWID, ' + group_by + ', COUNT(*) FROM ' + TABLE_NAME
+        query += ' GROUP BY ' + str(group_by)
+        if limit:
+            query += ' ORDER BY ROWID DESC LIMIT ' + str(limit)
+
+        try:
+            conn = sqlite3.connect(self.__file_path)
+            rows = conn.execute(query).fetchall()
```

Comment:
Connecting and executing are common among all the methods of this class. Factor out logic like this to a central method.